### ISSUE TITLE
implement of muon optimizer

### ISSUE THREAD

**Issue by pass-lin** [keras-team/keras#20952]
[Moun optimizer](https://github.com/KellerJordan/Muon) is an optimizer proposed by OpenAI that is stronger than AdamW. And it has been verified on the [Moonlight model](https://hf-mirror.com/moonshotai/Moonlight-16B-A3B-Instruct).
Has the Keras team implemented his plan? If not yet, I can submit a relevant PR.
If I were to provide the relevant PR, what should I pay attention to?

**Comment by pass-lin** [comment]
Hello, may I ask if I can submit a PR for the muon optimizer in Keras?

**Comment by hertschuh** [comment]
@pass-lin 

Would you be able to provide an implementation and example use as a code example on [keras.io](https://github.com/keras-team/keras-io/tree/master/examples) ?

Thanks!

**Comment by pass-lin** [comment]
> [@pass-lin](https://github.com/pass-lin)
> 
> Would you be able to provide an implementation and example use as a code example on [keras.io](https://github.com/keras-team/keras-io/tree/master/examples) ?
> 
> Thanks!

I'm happy to do that. But I want to ask, why isn't it a new feature of keras.optimizers but an example?

**Comment by hertschuh** [comment]
@pass-lin 

Actually, yes, please put it in `keras.optimizers`.

Thanks for the contribution!

**Comment by pass-lin** [comment]
> [@pass-lin](https://github.com/pass-lin)
> 
> Actually, yes, please put it in `keras.optimizers`.
> 
> Thanks for the contribution!
```
This optimizer should not be used for the embedding layer, the final fully connected layer,
    or any {0,1}-D parameters; those should all be optimized by a standard method (e.g., AdamW).
```
This is a warning from the author of moun, it seems that if you want to use moun, you need to use multiple optimizers at the same time. How can we implement this in Keras?

**Comment by hertschuh** [comment]
There are two ways to do this:
1. A custom `train_step` like https://keras.io/examples/keras_recipes/trainer_pattern/ At the part where you apply the gradients, you use different optimizers for different layers.
2. Via a multi-optimizer, i.e. an optimizer that dispatches to different optimizers based on some criteria. This doesn't exist in Keras 3 today. But I can work on this part.

**Comment by pass-lin** [comment]
> There are two ways to do this:
> 
> 1. A custom `train_step` like https://keras.io/examples/keras_recipes/trainer_pattern/ At the part where you apply the gradients, you use different optimizers for different layers.
> 2. Via a multi-optimizer, i.e. an optimizer that dispatches to different optimizers based on some criteria. This doesn't exist in Keras 3 today. But I can work on this part.

My idea is to have an adamw optimizer built into the mouns optimizer of keras. Choose one of the optimizers to execute according to the situation

**Comment by hertschuh** [comment]
> My idea is to have an adamw optimizer built into the mouns optimizer of keras. Choose one of the optimizers to execute according to the situation

Oh ok, that works too. You'll need to provide a way to say which layers use which optimizer.

One way to do it would be a regex on the layer and/or variable name. Similar to this: https://github.com/keras-team/keras/blob/master/keras/src/distribution/distribution_lib.py#L686-L718
